### PR TITLE
Remove destination with configuration splitting

### PR DESCRIPTION
**Pull request by ana-pantilie**
https://github.com/kframework/kore/pull/1300 should be merged before this.

---

###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`

### CODE REVIEW

File: kore/src/Kore/Step/Simplification/Exists.hs [hunk 1]
```diff
@@ -156,41 +167,73 @@ simplifyEvaluated variable simplified
...
-            matched <- Monad.Trans.lift $ matchesToVariableSubstitution
-                variable
-                normalized { Conditional.substitution = boundSubstitution }
-            if matched
-                then return normalized
-                    { Conditional.predicate = Predicate.makeTruePredicate_
-                    , Conditional.substitution = freeSubstitution
-                    }
-                else makeEvaluateBoundRight
+makeEvaluate variables original = do
```
Comment (ttuegel) [inline kore/src/Kore/Step/Simplification/Exists.hs]: Is the primary change here that we are simplifying multiple quantifiers at once? (The list of variables.)

File: kore/src/Kore/Strategies/Goal.hs [hunk 2]
```diff
@@ -731,22 +741,29 @@ allPathFollowupStep claims axioms =
...
-    pure $ makeRuleFromPatterns goal result destination
+    removal <- removalPredicate (destination right') configuration
+    let result = Conditional.andPredicate configuration removal
+    pure $ makeRuleFromPatterns goal result (Pattern.fromTermLike right')
   where
     configuration = getConfiguration goal
     configFreeVars = Pattern.freeVariables configuration
 
-    RulePattern { right, ensures } = toRulePattern goal
+    RulePattern { right } = toRulePattern goal
```
Comment (ttuegel) [inline kore/src/Kore/Strategies/Goal.hs]: Should we check that `ensures` is `\top`?

File: kore/src/Kore/Strategies/Goal.hs [hunk 3]
```diff
@@ -921,39 +938,88 @@ deriveSeq rules goal = errorBracket $ do
...
+                                Exists.makeEvaluate extraElemVariables remainderPattern
+                            return
+                                . Predicate.makeNotPredicate
+                                . Condition.toPredicate
+                                . Conditional.withoutTerm
+                                . OrPattern.toPattern
+                                $ evaluatedRemainder
+                _ -> error . show . Pretty.vsep $
+                    [ "Unifying the terms of the configuration and the\
+                      \ destination has unexpectedly produced more than one\
```
Comment (ttuegel) [inline kore/src/Kore/Strategies/Goal.hs]: Indentation should be a multiple of four.

File: kore/src/Kore/Strategies/Goal.hs [hunk 4]
```diff
@@ -921,39 +938,88 @@ deriveSeq rules goal = errorBracket $ do
...
+    destination
+    configuration
+  | isFunctionPattern configTerm
+  , isFunctionPattern destTerm
+  = do
+    unifiedConfigs <- simplifyTerm (mkAnd configTerm destTerm)
+    if OrPattern.isFalse unifiedConfigs
+        then return Predicate.makeTruePredicate_
+        else
+            case Foldable.toList unifiedConfigs of
```
Comment (ttuegel) [inline kore/src/Kore/Strategies/Goal.hs]:
```suggestion
            case OrPattern.toPatterns unifiedConfigs of
```

File: kore/src/Kore/Strategies/Goal.hs [hunk 5]
```diff
@@ -921,39 +938,88 @@ deriveSeq rules goal = errorBracket $ do
...
+                [substPattern] ->
+                    let extraNonElemVariables =
+                            remainderNonElemVariables configuration substPattern
+                        extraElemVariables =
+                            remainderElementVariables configuration substPattern
+                    in if not (null extraNonElemVariables)
+                        then
+                            error . show . Pretty.vsep $
+                                "Cannot quantify non-element variables: "
+                                : fmap (Pretty.indent 4 . unparse) extraNonElemVariables
```
Comment (ttuegel) [inline kore/src/Kore/Strategies/Goal.hs]: Let's factor this branch out into a function which returns `extraElemVariables` or throws an error if there are non-element variables. Then, we only need to write the "happy path" here.

File: kore/test/Test/Kore/Strategies/AllPath/Verification.hs [hunk 6]
```diff
@@ -298,7 +303,15 @@ simpleClaim
     -> TermLike Variable
     -> AllPathRule Variable
 simpleClaim left right =
-    AllPathRule . getRewriteRule $ simpleRewrite left right
+    AllPathRule
+    $ RulePattern
```
Comment (ttuegel) [inline kore/test/Test/Kore/Strategies/AllPath/Verification.hs]: Record brackets bind stronger than application:

```suggestion
    RulePattern
```

File: kore/test/Test/Kore/Strategies/OnePath/Verification.hs [hunk 7]
```diff
@@ -345,7 +344,15 @@ simpleClaim
     -> TermLike Variable
     -> OnePathRule Variable
 simpleClaim left right =
-    OnePathRule . getRewriteRule $ simpleRewrite left right
+    OnePathRule
+    $ RulePattern
```
Comment (ttuegel) [inline kore/test/Test/Kore/Strategies/OnePath/Verification.hs]:
```suggestion
    RulePattern
```

File: kore/src/Kore/Step/Simplification/Exists.hs (hunk 1 above)
Comment (ana-pantilie) [reply kore/src/Kore/Step/Simplification/Exists.hs]: Yes, that should be it. I updated the documentation here as @traiansf suggested.

File: kore/src/Kore/Step/Simplification/Exists.hs [hunk 8]
```diff
@@ -156,41 +167,76 @@ simplifyEvaluated variable simplified
...
+    let sortedVariables = sortBy substVariablesFirst variables
```
Comment (virgil-serbanuta) [inline kore/src/Kore/Step/Simplification/Exists.hs]: Shouldn't we resimplify the predicate and then resort the variables after substituting the ones that we can? Alternately, we could call `simplifyCondition` before sorting, which may extract some substitutions that are "pendind" inside the predicate, e.g. a postponed map lookup (I don't remember if we still have those).

Comment (ana-pantilie) [reply kore/src/Kore/Step/Simplification/Exists.hs]: @virgil-serbanuta Does it matter if we call `simplifyCondition` before or after sorting the variables? We don't use the variables when simplifying the condition, and the (newly-formed) condition always gets simplified at the beginning of `makeEvaluateWorker`. Have I misunderstood what you're saying?

Comment (virgil-serbanuta) [reply kore/src/Kore/Step/Simplification/Exists.hs]: My question is something like this: can the `simplifyCondition original'` call in `makeEvaluateWorker` add more variables to the substitution? If yes, should that change the sorting order?

As an example, if our `original` condition is `C(x)=f(y) /\ [y=5]`, where `C` is a constructor, `f` is a function and `f(5)=C(A)`, then should we simplify that to `[x=A, y=5]` before sorting, or we should consider `x` as not being a substituted variable?

Now, the above will not be a valid example after https://github.com/kframework/kore/pull/1320 , but you can ask the same about `C(A)=f(x,y) and [y=5]` where `f(x, 5)=C(x)`.

Comment (ana-pantilie) [reply kore/src/Kore/Step/Simplification/Exists.hs]: Oh, I see. Thank you for the explanation! I'll try to fix this in a follow-up PR.

**Pull request merged**